### ISSUE TITLE
Support for color difference as given by the CIEDE 2000 standard

### ISSUE THREAD

**Issue by Mengda-Li** [kornia/kornia#3062]
## 🚀 Feature
Could you add implementation of deltaE_ciede(2000) distance to measure the color difference as given by the CIEDE 2000 standard?
[https://scikit-image.org/docs/stable/api/skimage.color.html#skimage.color.deltaE_ciede2000](https://scikit-image.org/docs/stable/api/skimage.color.html#skimage.color.deltaE_ciede2000)

## Motivation

It is an important perceptual distance to measure the color difference in human eyes.

## Pitch

Such as translate the `numpy` inplementation in `skimage` to `torch`
[https://github.com/scikit-image/scikit-image/blob/v0.24.0/skimage/color/delta_e.py#L152-L283](https://github.com/scikit-image/scikit-image/blob/v0.24.0/skimage/color/delta_e.py#L152-L283)

```python
def deltaE_ciede2000(lab1, lab2, kL=1, kC=1, kH=1, *, channel_axis=-1):
    """Color difference as given by the CIEDE 2000 standard.

    CIEDE 2000 is a major revision of CIDE94.  The perceptual calibration is
    largely based on experience with automotive paint on smooth surfaces.

    Parameters
    ----------
    lab1 : array_like
        reference color (Lab colorspace)
    lab2 : array_like
        comparison color (Lab colorspace)
    kL : float (range), optional
        lightness scale factor, 1 for "acceptably close"; 2 for "imperceptible"
        see deltaE_cmc
    kC : float (range), optional
        chroma scale factor, usually 1
    kH : float (range), optional
        hue scale factor, usually 1
    channel_axis : int, optional
        This parameter indicates which axis of the arrays corresponds to
        channels.

        .. versionadded:: 0.19
           ``channel_axis`` was added in 0.19.

    Returns
    -------
    deltaE : array_like
        The distance between `lab1` and `lab2`

    Notes
    -----
    CIEDE 2000 assumes parametric weighting factors for the lightness, chroma,
    and hue (`kL`, `kC`, `kH` respectively).  These default to 1.

    References
    ----------
    .. [1] https://en.wikipedia.org/wiki/Color_difference
    .. [2] http://www.ece.rochester.edu/~gsharma/ciede2000/ciede2000noteCRNA.pdf
           :DOI:`10.1364/AO.33.008069`
    .. [3] M. Melgosa, J. Quesada, and E. Hita, "Uniformity of some recent
           color metrics tested with an accurate color-difference tolerance
           dataset," Appl. Opt. 33, 8069-8077 (1994).
    """
    lab1, lab2 = _float_inputs(lab1, lab2, allow_float32=True)

    channel_axis = channel_axis % lab1.ndim
    unroll = False
    if lab1.ndim == 1 and lab2.ndim == 1:
        unroll = True
        if lab1.ndim == 1:
            lab1 = lab1[None, :]
        if lab2.ndim == 1:
            lab2 = lab2[None, :]
        channel_axis += 1
    L1, a1, b1 = np.moveaxis(lab1, source=channel_axis, destination=0)[:3]
    L2, a2, b2 = np.moveaxis(lab2, source=channel_axis, destination=0)[:3]

    # distort `a` based on average chroma
    # then convert to lch coordinates from distorted `a`
    # all subsequence calculations are in the new coordinates
    # (often denoted "prime" in the literature)
    Cbar = 0.5 * (np.hypot(a1, b1) + np.hypot(a2, b2))
    c7 = Cbar**7
    G = 0.5 * (1 - np.sqrt(c7 / (c7 + 25**7)))
    scale = 1 + G
    C1, h1 = _cart2polar_2pi(a1 * scale, b1)
    C2, h2 = _cart2polar_2pi(a2 * scale, b2)
    # recall that c, h are polar coordinates.  c==r, h==theta

    # cide2000 has four terms to delta_e:
    # 1) Luminance term
    # 2) Hue term
    # 3) Chroma term
    # 4) hue Rotation term

    # lightness term
    Lbar = 0.5 * (L1 + L2)
    tmp = (Lbar - 50) ** 2
    SL = 1 + 0.015 * tmp / np.sqrt(20 + tmp)
    L_term = (L2 - L1) / (kL * SL)

    # chroma term
    Cbar = 0.5 * (C1 + C2)  # new coordinates
    SC = 1 + 0.045 * Cbar
    C_term = (C2 - C1) / (kC * SC)

    # hue term
    h_diff = h2 - h1
    h_sum = h1 + h2
    CC = C1 * C2

    dH = h_diff.copy()
    dH[h_diff > np.pi] -= 2 * np.pi
    dH[h_diff < -np.pi] += 2 * np.pi
    dH[CC == 0.0] = 0.0  # if r == 0, dtheta == 0
    dH_term = 2 * np.sqrt(CC) * np.sin(dH / 2)

    Hbar = h_sum.copy()
    mask = np.logical_and(CC != 0.0, np.abs(h_diff) > np.pi)
    Hbar[mask * (h_sum < 2 * np.pi)] += 2 * np.pi
    Hbar[mask * (h_sum >= 2 * np.pi)] -= 2 * np.pi
    Hbar[CC == 0.0] *= 2
    Hbar *= 0.5

    T = (
        1
        - 0.17 * np.cos(Hbar - np.deg2rad(30))
        + 0.24 * np.cos(2 * Hbar)
        + 0.32 * np.cos(3 * Hbar + np.deg2rad(6))
        - 0.20 * np.cos(4 * Hbar - np.deg2rad(63))
    )
    SH = 1 + 0.015 * Cbar * T

    H_term = dH_term / (kH * SH)

    # hue rotation
    c7 = Cbar**7
    Rc = 2 * np.sqrt(c7 / (c7 + 25**7))
    dtheta = np.deg2rad(30) * np.exp(-(((np.rad2deg(Hbar) - 275) / 25) ** 2))
    R_term = -np.sin(2 * dtheta) * Rc * C_term * H_term

    # put it all together
    dE2 = L_term**2
    dE2 += C_term**2
    dE2 += H_term**2
    dE2 += R_term
    ans = np.sqrt(np.maximum(dE2, 0))
    if unroll:
        ans = ans[0]
    return ans
```

**Comment by krahnikblis** [comment]
here, i was doing this stuff anyway when i was searching for Kornia maybe already having these things (glad it didn't; i'd feel i wasted a lot of time)

these are refactored from sk-image for the pytroch, and are needed within the distance function:
```python

class channel_as_last_axis:
	"""Decorator for automatically making channels axis last for all arrays.

	This decorator reorders axes for compatibility with functions that only
	support channels along the last axis. After the function call is complete
	the channels axis is restored back to its original position.

	Parameters
	----------
	channel_arg_positions : tuple of int, optional
		Positional arguments at the positions specified in this tuple are
		assumed to be multichannel arrays. The default is to assume only the
		first argument to the function is a multichannel array.
	channel_kwarg_names : tuple of str, optional
		A tuple containing the names of any keyword arguments corresponding to
		multichannel arrays.
	multichannel_output : bool, optional
		A boolean that should be True if the output of the function is not a
		multichannel array and False otherwise. This decorator does not
		currently support the general case of functions with multiple outputs
		where some or all are multichannel.

	"""

	def __init__(
		self,
		channel_arg_positions=(0,),
		channel_kwarg_names=(),
		multichannel_output=True,
		):
		self.arg_positions = set(channel_arg_positions)
		self.kwarg_names = set(channel_kwarg_names)
		self.multichannel_output = multichannel_output

	def _torch_isscalar(self, x):
		"""
		Check if the input is a scalar.

		Parameters
		----------
		x : Any
			Input to check for scalar property.

		Returns
		-------
		bool
			True if the input is a scalar, False otherwise.
		"""
		return isinstance(x, (int, float, complex, torch.Tensor)) and (
			not isinstance(x, torch.Tensor) or x.ndim == 0
			)

	def __call__(self, func):
		@functools.wraps(func)
		def fixed_func(*args, **kwargs):
			channel_axis = kwargs.get('channel_axis', None)

			if channel_axis is None:
				return func(*args, **kwargs)

			# TODO: convert scalars to a tuple in anticipation of eventually
			#       supporting a tuple of channel axes. Right now, only an
			#       integer or a single-element tuple is supported, though.
			if self._torch_isscalar(channel_axis):
				channel_axis = (channel_axis,)
			if len(channel_axis) > 1:
				raise ValueError("only a single channel axis is currently supported")

			if channel_axis == (-1,) or channel_axis == -1:
				return func(*args, **kwargs)

			if self.arg_positions:
				new_args = []
				for pos, arg in enumerate(args):
					if pos in self.arg_positions:
						new_args.append(torch.movedim(arg, channel_axis[0], -1))
					else:
						new_args.append(arg)
				new_args = tuple(new_args)
			else:
				new_args = args

			for name in self.kwarg_names:
				kwargs[name] = torch.movedim(kwargs[name], channel_axis[0], -1)

			# now that we have moved the channels axis to the last position,
			# change the channel_axis argument to -1
			kwargs["channel_axis"] = -1

			# Call the function with the fixed arguments
			out = func(*new_args, **kwargs)
			if self.multichannel_output:
				out = torch.movedim(out, -1, channel_axis[0])
			return out

		return fixed_func
		
def _cart2polar_2pi(x, y):
	"""convert cartesian coordinates to polar (uses non-standard theta range!)

	NON-STANDARD RANGE! Maps to ``(0, 2*pi)`` rather than usual ``(-pi, +pi)``
	"""
	r, t = torch.hypot(x, y), torch.arctan2(y, x)
	t += torch.where(t < 0.0, 2 * torch.pi, 0)
	return r, t

@channel_as_last_axis()
def lab2lch(lab, *, channel_axis=-1):
	"""Convert image in CIE-LAB to CIE-LCh color space.

	CIE-LCh is the cylindrical representation of the CIE-LAB (Cartesian) color
	space.

	Parameters
	----------
	lab : (..., C=3, ...) array_like
		The input image in CIE-LAB color space.
		Unless `channel_axis` is set, the final dimension denotes the CIE-LAB
		channels.
		The L* values range from 0 to 100;
		the a* and b* values range from -128 to 127.
	channel_axis : int, optional
		This parameter indicates which axis of the array corresponds to
		channels.

		.. versionadded:: 0.19
		``channel_axis`` was added in 0.19.

	Returns
	-------
	out : (..., C=3, ...) ndarray
		The image in CIE-LCh color space, of same shape as input.

	Raises
	------
	ValueError
		If `lab` does not have at least 3 channels (i.e., L*, a*, and b*).

	Notes
	-----
	The h channel (i.e., hue) is expressed as an angle in range ``(0, 2*pi)``.

	See Also
	--------
	lch2lab

	References
	----------
	.. [1] http://www.easyrgb.com/en/math.php
	.. [2] https://en.wikipedia.org/wiki/CIELAB_color_space
	.. [3] https://en.wikipedia.org/wiki/HCL_color_space

	"""
	lch = lab.clone()

	a, b = lch[..., 1], lch[..., 2]
	lch[..., 1], lch[..., 2] = _cart2polar_2pi(a, b)
	return lch
```

and here's the requested distance function refactored:
```python

def deltaE_ciede2000(lab1, lab2, kL=1, kC=1, kH=1, *, channel_axis=-1):
	"""Color difference as given by the CIEDE 2000 standard.

	CIEDE 2000 is a major revision of CIDE94.  The perceptual calibration is
	largely based on experience with automotive paint on smooth surfaces.

	Parameters
	----------
	lab1 : array_like
		reference color (Lab colorspace)
	lab2 : array_like
		comparison color (Lab colorspace)
	kL : float (range), optional
		lightness scale factor, 1 for "acceptably close"; 2 for "imperceptible"
		see deltaE_cmc
	kC : float (range), optional
		chroma scale factor, usually 1
	kH : float (range), optional
		hue scale factor, usually 1

	Returns
	-------
	deltaE : array_like
		The distance between `lab1` and `lab2`

	Notes
	-----
	CIEDE 2000 assumes parametric weighting factors for the lightness, chroma,
	and hue (`kL`, `kC`, `kH` respectively).  These default to 1.

	References
	----------
	.. [1] https://en.wikipedia.org/wiki/Color_difference
	.. [2] http://www.ece.rochester.edu/~gsharma/ciede2000/ciede2000noteCRNA.pdf
		:DOI:`10.1364/AO.33.008069`
	.. [3] M. Melgosa, J. Quesada, and E. Hita, "Uniformity of some recent
		color metrics tested with an accurate color-difference tolerance
		dataset," Appl. Opt. 33, 8069-8077 (1994).
	"""

	channel_axis = channel_axis % lab1.ndim
	unroll = False
	if lab1.ndim == 1 and lab2.ndim == 1:
		unroll = True
		if lab1.ndim == 1:
			dlab1 = lab1.clone()[None, :]
		if lab2.ndim == 1:
			dlab2 = lab2.clone()[None, :]
		channel_axis += 1
	else:
		dlab1 = lab1.clone()
		dlab2 = lab2.clone()
	L1, a1, b1 = torch.movedim(dlab1, source=channel_axis, destination=0)[:3]
	L2, a2, b2 = torch.movedim(dlab2, source=channel_axis, destination=0)[:3]

	# distort `a` based on average chroma
	# then convert to lch coordinates from distorted `a`
	# all subsequence calculations are in the new coordinates
	# (often denoted "prime" in the literature)
	Cbar = 0.5 * (torch.hypot(a1, b1) + torch.hypot(a2, b2))
	c7 = Cbar**7
	G = 0.5 * (1 - torch.sqrt(c7 / (c7 + 25**7)))
	scale = 1 + G
	C1, h1 = _cart2polar_2pi(a1 * scale, b1)
	C2, h2 = _cart2polar_2pi(a2 * scale, b2)
	# recall that c, h are polar coordinates.  c==r, h==theta

	# cide2000 has four terms to delta_e:
	# 1) Luminance term
	# 2) Hue term
	# 3) Chroma term
	# 4) hue Rotation term

	# lightness term
	Lbar = 0.5 * (L1 + L2)
	tmp = (Lbar - 50) ** 2
	SL = 1 + 0.015 * tmp / torch.sqrt(20 + tmp)
	L_term = (L2 - L1) / (kL * SL)

	# chroma term
	Cbar = 0.5 * (C1 + C2)  # new coordinates
	SC = 1 + 0.045 * Cbar
	C_term = (C2 - C1) / (kC * SC)

	# hue term
	h_diff = h2 - h1
	h_sum = h1 + h2
	CC = C1 * C2

	dH = h_diff.clone()
	dH[h_diff > torch.pi] -= 2 * torch.pi
	dH[h_diff < -torch.pi] += 2 * torch.pi
	dH[CC == 0.0] = 0.0  # if r == 0, dtheta == 0
	dH_term = 2 * torch.sqrt(CC) * torch.sin(dH / 2)

	Hbar = h_sum.clone()
	mask = torch.logical_and(CC != 0.0, torch.abs(h_diff) > torch.pi)
	Hbar[mask * (h_sum < 2 * torch.pi)] += 2 * torch.pi
	Hbar[mask * (h_sum >= 2 * torch.pi)] -= 2 * torch.pi
	Hbar[CC == 0.0] *= 2
	Hbar *= 0.5

	T = (
		1
		- 0.17 * torch.cos(Hbar - torch.deg2rad(torch.asarray(30)))
		+ 0.24 * torch.cos(2 * Hbar)
		+ 0.32 * torch.cos(3 * Hbar + torch.deg2rad(torch.asarray(6)))
		- 0.20 * torch.cos(4 * Hbar - torch.deg2rad(torch.asarray(63)))
		)
	SH = 1 + 0.015 * Cbar * T

	H_term = dH_term / (kH * SH)

	# hue rotation
	c7 = Cbar**7
	Rc = 2 * torch.sqrt(c7 / (c7 + 25**7))
	dtheta = torch.deg2rad(torch.asarray(30)) * torch.exp(-(((torch.rad2deg(Hbar) - 275) / 25) ** 2))
	R_term = -torch.sin(2 * dtheta) * Rc * C_term * H_term

	# put it all together
	dE2 = L_term**2
	dE2 += C_term**2
	dE2 += H_term**2
	dE2 += R_term
	ans = torch.sqrt(torch.maximum(dE2, torch.asarray(0)))
	if unroll:
		ans = ans[0]
	return ans
```

from what i can tell, Kornia puts channel in the -3 dimension? so either these would need a little editing to hard-code for that, or just implement the channel decorator function as well as i pasted in (ganked from sk-image) and be sure to use it if channel is in -3

**Comment by edgarriba** [comment]
i would go for adapting the algorithm to the 3rd channel, otherwise too much overhead permuting axis and accessors
